### PR TITLE
#428: Filter null value for custom field update actions

### PR DESCRIPTION
**Pull request by JudeNiroshan**
#### Description

I want to be able to set null for custom price field attribute value:

Set null as value without failures
Should not generate 2 update actions when new value is null
...
so that:
sync-java will create only remove update action as the value is null
should not fail the sync process because of trying to update resource which do not exists

[https://github.com/commercetools/commercetools-sync-java/issues/428](https://github.com/commercetools/commercetools-sync-java/issues/428)
#### Todo

- Tests
    - [x] Unit 
    - [x] Integration
- [x] Documentation
- [x] Add Release Notes entry.

### CODE REVIEW

**Comment by ahmetoz** [comment]
Also, it could be nice if you could add this to the Release notes as a bug fix.

File: src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java [hunk 1]
```diff
@@ -342,6 +342,7 @@
         @Nonnull final Function<T, String> updateIdGetter) {
 
         return newCustomFields.keySet().stream()
+                              .filter(newCustomFieldName -> Objects.nonNull(newCustomFields.get(newCustomFieldName)))
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: it would be nice if you update the Javadoc too, something like this: 
`<p>Note: Null custom fields are filtered out.`

Comment (JudeNiroshan) [reply src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: aec6bf09c83bd97530d157cf40ff923842896a37

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java [hunk 2]
```diff
@@ -586,6 +591,220 @@ void withMixedCasesOfPriceMatches_ShouldBuildActions() {
...
+        assertThat(productProjection).isNotNull();
+
+        List<Price> priceList = productProjection.getMasterVariant().getPrices();
+        for (Price price : priceList) {
+            Set<String> fieldAsStringSet = price.getCustom().getFieldAsStringSet(SET_CUSTOM_FIELD_NAME);
+            assertThat(fieldAsStringSet).contains("something");
+        }
+    }
+
+    //todo Need to enable @Test when the fix is released from jvm-sdk
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: https://github.com/commercetools/commercetools-sync-java/issues/430

35cce6653bbc3e349f96d76da6ff65a02e715692

File: docs/RELEASE_NOTES.md [hunk 3]
```diff
@@ -34,7 +34,7 @@
 [Jar](https://bintray.com/commercetools/maven/commercetools-sync-java/1.5.0)
 
 - 🐞 **Bug Fixes** (1)
-    - **Commons** -  Fixed a bug in the `CustomUpdateActionUtils` which caused to generate unnecessary update actions in `buildNewOrModifiedCustomFieldsUpdateActions`. [#428](https://github.com/commercetools/commercetools-sync-java/issues/428)
+    - **Commons** -  Fixed a bug in the `CustomUpdateActionUtils#buildNewOrModifiedCustomFieldsUpdateActions` which caused to generate unnecessary update actions for null custom field values. #428
```
Comment (ahmetoz) [inline docs/RELEASE_NOTES.md]: it seems you forgot the add link on markdown for `#428`

Comment (JudeNiroshan) [reply docs/RELEASE_NOTES.md]: 4c956b8c1c00bcea4a7136330f6d2fe2fdb7e5ef

**Comment by JudeNiroshan** [comment]
> I think for here, we need to keep it as commented and remove the comment on the release preparation PR. Check the old PR (#423), how we are releasing the new version of the java-sync.

36ef5c9a1892929cbc683220e11ab266327863f4
I only commented the header. Do you think we should comment the entire release note section?

**Comment by ahmetoz** [comment]
> I only commented the header. Do you think we should comment the entire release note section?

Entire header.

**Comment by JudeNiroshan** [comment]
> Good job here 👍 💯
> Note: It would be great if you add a unit test for the null check _if you have time_.

90a3cc9c358c4cb54c409a4aaa797e9c25f1388c

**Comment by JudeNiroshan** [comment]
> > I only commented the header. Do you think we should comment the entire release note section?
> 
> Entire release notes.

dc9da5a5cdd131ce77439f99c6df16c9003d6b1d

**Comment by ahmetoz** [comment]
don't forget 🤔  https://github.com/commercetools/commercetools-sync-java/pull/429#commitcomment-34806346

File: docs/RELEASE_NOTES.md [hunk 4]
```diff
@@ -34,6 +34,12 @@
 [Javadoc](https://commercetools.github.io/commercetools-sync-java/v/1.5.0/) | 
 [Jar](https://bintray.com/commercetools/maven/commercetools-sync-java/1.5.0)
 
+- 🐞 **Bug Fixes** (1)
+    - **Commons** -  Fixed a bug in the `CustomUpdateActionUtils#buildNewOrModifiedCustomFieldsUpdateActions` which caused to generate unnecessary update actions for null custom field values. [#428](https://github.com/commercetools/commercetools-sync-java/issues/428)
+
+- 🛠️ **Enhancements** (1)
+    - **Commercetools-jvm-sdk** - Upgraded to version 1.46.0 which includes a PriceCustomField serialization bug [#430](https://github.com/commercetools/commercetools-sync-java/issues/430)
```
Comment (heshamMassoud) [inline docs/RELEASE_NOTES.md]:
```suggestion
    - **Commons** - Upgraded commercetools-jvm-sdk to version [1.46.0](http://commercetools.github.io/commercetools-jvm-sdk/apidocs/io/sphere/sdk/meta/ReleaseNotes.html#v1_46_0) which includes a PriceCustomField serialization bug. [#430](https://github.com/commercetools/commercetools-sync-java/issues/430)
```

File: docs/RELEASE_NOTES.md [hunk 5]
```diff
@@ -34,6 +34,12 @@
 [Javadoc](https://commercetools.github.io/commercetools-sync-java/v/1.5.0/) | 
 [Jar](https://bintray.com/commercetools/maven/commercetools-sync-java/1.5.0)
 
+- 🐞 **Bug Fixes** (1)
+    - **Commons** -  Fixed a bug in the `CustomUpdateActionUtils#buildNewOrModifiedCustomFieldsUpdateActions` which caused to generate unnecessary update actions for null custom field values. [#428](https://github.com/commercetools/commercetools-sync-java/issues/428)
```
Comment (heshamMassoud) [inline docs/RELEASE_NOTES.md]:
```suggestion
    - **Commons** -  Fixed a bug in the `CustomUpdateActionUtils#buildNewOrModifiedCustomFieldsUpdateActions` which generated unnecessary update actions for null custom field values. [#428](https://github.com/commercetools/commercetools-sync-java/issues/428)
```

File: src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java [hunk 6]
```diff
@@ -319,6 +319,8 @@
      * value or the addition of a new one. It returns a {@link List}&lt;{@link UpdateAction}&gt; as a result.
      * If no update action is needed an empty {@link List}&lt;{@link UpdateAction}&gt; is returned.
      *
+     * <p>Note: Null value custom fields are filtered out.
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: why do we filter out null values? Why do we skip building update actions when the new value is `null`?

Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: ok, now I understand why. Because the other method `buildRemovedCustomFieldsUpdateActions` takes care of that. Would only suggest:
```suggestion
     * <p>Note: Null value custom fields are filtered out. In other words, no update actions would be built for fields with null values.
```

File: src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java [hunk 7]
```diff
@@ -546,6 +546,22 @@ void buildNewOrModifiedCustomFieldsUpdateActions_WithNoNewOrModifiedCustomFields
         assertThat(updateActions).isEmpty();
     }
 
+    @Test
+    void buildNewOrModifiedCustomFieldsUpdateActions_WithNullCustomFields_ShouldNotBuildActions() {
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java]: nice that you added this unit test. Although, I don't like that we test these protected methods. But it was a problem that was already there. I guess we should make these utils private and test them indirectly by testing the caller method. But this is something that we should not take care about right now I guess.

I will create an issue for it.

File: src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java [hunk 8]
```diff
@@ -546,6 +546,22 @@ void buildNewOrModifiedCustomFieldsUpdateActions_WithNoNewOrModifiedCustomFields
...
+        final Map<String, JsonNode> newCustomFields = new HashMap<>();
+        newCustomFields.put("invisibleInShop", null);
+
+        final List<UpdateAction<Product>> updateActions = buildNewOrModifiedCustomFieldsUpdateActions(oldCustomFields,
+            newCustomFields, mock(Price.class), new PriceCustomActionBuilder(), 1, Price::getId);
+
+        assertThat(updateActions).isNotNull();
+        assertThat(updateActions).isEmpty();
+    }
+
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java]: I would also add these unit tests:

```java
    @Test
    void buildSetCustomFieldsUpdateActions_WithNullAsValue_ShouldCorrectlyBuildAction() {
        // preparation
        final Map<String, JsonNode> oldCustomFieldsMap = new HashMap<>();
        oldCustomFieldsMap.put("setOfBooleans", JsonNodeFactory
            .instance
            .arrayNode()
            .add(JsonNodeFactory.instance.booleanNode(false)));

        final CustomFields oldCustomFields = mock(CustomFields.class);
        when(oldCustomFields.getFieldsJsonMap()).thenReturn(oldCustomFieldsMap);

        final Asset oldAsset = mock(Asset.class);
        when(oldAsset.getCustom()).thenReturn(oldCustomFields);


        final Map<String, JsonNode> newCustomFieldsMap = new HashMap<>();
        newCustomFieldsMap.put("setOfBooleans", null);


        // test
        final List<UpdateAction<Product>> updateActions =
            buildSetCustomFieldsUpdateActions(oldCustomFieldsMap, newCustomFieldsMap, mock(Asset.class), new AssetCustomActionBuilder(), 1, Asset::getId);

        // assertion
        assertThat(updateActions)
            .containsExactly(SetAssetCustomField
                .ofVariantIdUsingJsonAndAssetKey(1, oldAsset.getKey(), "setOfBooleans", null, true));
    }

    @Test
    void buildSetCustomFieldsUpdateActions_WithNullJsonNodeAsValue_ShouldCorrectlyBuildAction() {
        // preparation
        final Map<String, JsonNode> oldCustomFieldsMap = new HashMap<>();
        oldCustomFieldsMap.put("setOfBooleans", JsonNodeFactory
            .instance
            .arrayNode()
            .add(JsonNodeFactory.instance.booleanNode(false)));

        final CustomFields oldCustomFields = mock(CustomFields.class);
        when(oldCustomFields.getFieldsJsonMap()).thenReturn(oldCustomFieldsMap);

        final Asset oldAsset = mock(Asset.class);
        when(oldAsset.getCustom()).thenReturn(oldCustomFields);


        final Map<String, JsonNode> newCustomFieldsMap = new HashMap<>();
        newCustomFieldsMap.put("setOfBooleans", JsonNodeFactory.instance.nullNode());


        // test
        final List<UpdateAction<Product>> updateActions =
            buildSetCustomFieldsUpdateActions(oldCustomFieldsMap, newCustomFieldsMap, mock(Asset.class), new AssetCustomActionBuilder(), 1, Asset::getId);

        // assertion
        assertThat(updateActions)
            .containsExactly(SetAssetCustomField
                .ofVariantIdUsingJsonAndAssetKey(1, oldAsset.getKey(), "setOfBooleans", JsonNodeFactory.instance.nullNode(), true));
    }
```

File: src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java (hunk 7 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java]: https://github.com/commercetools/commercetools-sync-java/issues/434

File: src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java [hunk 9]
```diff
@@ -546,6 +546,22 @@ void buildNewOrModifiedCustomFieldsUpdateActions_WithNoNewOrModifiedCustomFields
...
+        oldCustomFields.put("invisibleInShop", JsonNodeFactory.instance.booleanNode(true));
+        oldCustomFields.put("backgroundColor", JsonNodeFactory.instance.objectNode().put("de", "rot").put("en", "red"));
+
+        final Map<String, JsonNode> newCustomFields = new HashMap<>();
+        newCustomFields.put("invisibleInShop", null);
+
+        final List<UpdateAction<Product>> updateActions = buildNewOrModifiedCustomFieldsUpdateActions(oldCustomFields,
+            newCustomFields, mock(Price.class), new PriceCustomActionBuilder(), 1, Price::getId);
+
+        assertThat(updateActions).isNotNull();
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java]: this assertion is not needed. It is already implicitly asserted in the following `isEmpty` assertion.

File: src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java (hunk 8 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java]: I would also add these 2 unit tests in `ProductVariantAssetUpdateActionUtilsTest`:


```java
    @Test
    void buildCustomUpdateActions_WithNullValue_ShouldCorrectlyBuildAction() {
        // preparation
        final Map<String, JsonNode> oldCustomFieldsMap = new HashMap<>();
        oldCustomFieldsMap.put("invisibleInShop", JsonNodeFactory.instance.booleanNode(true));
        oldCustomFieldsMap.put("setOfBooleans", JsonNodeFactory
            .instance
            .arrayNode()
            .add(JsonNodeFactory.instance.booleanNode(false)));

        final Map<String, JsonNode> newCustomFieldsMap = new HashMap<>();
        newCustomFieldsMap.put("invisibleInShop", JsonNodeFactory.instance.booleanNode(true));
        newCustomFieldsMap.put("setOfBooleans", null);


        final CustomFields oldCustomFields = mock(CustomFields.class);
        final String typeId = UUID.randomUUID().toString();
        when(oldCustomFields.getType()).thenReturn(Type.referenceOfId(typeId));
        when(oldCustomFields.getFieldsJsonMap()).thenReturn(oldCustomFieldsMap);

        final CustomFieldsDraft newCustomFieldsDraft =
            CustomFieldsDraft.ofTypeIdAndJson(typeId, newCustomFieldsMap);

        final Asset oldAsset = mock(Asset.class);
        when(oldAsset.getCustom()).thenReturn(oldCustomFields);

        final AssetDraft newAssetDraft = AssetDraftBuilder.of(emptyList(), empty())
                                                          .custom(newCustomFieldsDraft)
                                                          .build();

        final List<String> errors = new ArrayList<>();

        final ProductSyncOptions syncOptions = ProductSyncOptionsBuilder.of(mock(SphereClient.class))
                                                                        .errorCallback((errorMessage, throwable) ->
                                                                            errors.add(errorMessage))
                                                                        .build();
        // test
        final List<UpdateAction<Product>> updateActions =
            buildCustomUpdateActions(1, oldAsset, newAssetDraft, syncOptions);

        // assertion
        assertThat(errors).isEmpty();
        assertThat(updateActions)
            .containsExactly(SetAssetCustomField
                .ofVariantIdUsingJsonAndAssetKey(1, oldAsset.getKey(), "setOfBooleans", null, true));
    }

    @Test
    void buildCustomUpdateActions_WithNullJsonNodeValue_ShouldCorrectlyBuildAction() {
        // preparation
        final Map<String, JsonNode> oldCustomFieldsMap = new HashMap<>();
        oldCustomFieldsMap.put("setOfBooleans", JsonNodeFactory
            .instance
            .arrayNode()
            .add(JsonNodeFactory.instance.booleanNode(false)));

        final Map<String, JsonNode> newCustomFieldsMap = new HashMap<>();
        newCustomFieldsMap.put("setOfBooleans", JsonNodeFactory.instance.nullNode());


        final CustomFields oldCustomFields = mock(CustomFields.class);
        final String typeId = UUID.randomUUID().toString();
        when(oldCustomFields.getType()).thenReturn(Type.referenceOfId(typeId));
        when(oldCustomFields.getFieldsJsonMap()).thenReturn(oldCustomFieldsMap);

        final CustomFieldsDraft newCustomFieldsDraft =
            CustomFieldsDraft.ofTypeIdAndJson(typeId, newCustomFieldsMap);

        final Asset oldAsset = mock(Asset.class);
        when(oldAsset.getCustom()).thenReturn(oldCustomFields);

        final AssetDraft newAssetDraft = AssetDraftBuilder.of(emptyList(), empty())
                                                          .custom(newCustomFieldsDraft)
                                                          .build();

        final List<String> errors = new ArrayList<>();

        final ProductSyncOptions syncOptions = ProductSyncOptionsBuilder.of(mock(SphereClient.class))
                                                                        .errorCallback((errorMessage, throwable) ->
                                                                            errors.add(errorMessage))
                                                                        .build();
        // test
        final List<UpdateAction<Product>> updateActions =
            buildCustomUpdateActions(1, oldAsset, newAssetDraft, syncOptions);

        // assertion
        assertThat(errors).isEmpty();
        assertThat(updateActions)
            .containsExactly(SetAssetCustomField
                .ofVariantIdUsingJsonAndAssetKey(1, oldAsset.getKey(), "setOfBooleans", JsonNodeFactory.instance.nullNode(), true));
    }
```

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java [hunk 10]
```diff
@@ -586,6 +591,220 @@ void withMixedCasesOfPriceMatches_ShouldBuildActions() {
...
+
+        List<Price> priceList = productProjection.getMasterVariant().getPrices();
+        for (Price price : priceList) {
+            Set<String> fieldAsStringSet = price.getCustom().getFieldAsStringSet(SET_CUSTOM_FIELD_NAME);
+            assertThat(fieldAsStringSet).containsOnly("new something");
+        }
+    }
+
+    @Test
+    void withNullPriceCustomFieldSet_ForEmptyPriceCustomField_ShouldDoNothing() {
```
Comment (heshamMassoud) [inline src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: why should it do nothing? 

I mean empty set is different than a `null` value right?

Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: and actually as I see in your assertion it does something -> update

Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: I think we should an integration test for also null json node as value: 

```
JsonNodeFactory.instance.nullNode();
```

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java [hunk 11]
```diff
@@ -586,6 +591,220 @@ void withMixedCasesOfPriceMatches_ShouldBuildActions() {
         assertPricesAreEqual(productProjection.getMasterVariant().getPrices(), newPricesWithReferenceIds);
     }
 
+    @Test
+    void withNonEmptyPriceCustomFieldSet_ForEmptyPriceCustomField_ShouldCreateCustomFieldValue() {
```
Comment (heshamMassoud) [inline src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: Please follow the test naming convention: 

`methodUnderTest_testCondition_expectedResult`, in this case:


`sync_WithEmptyOldCustomFieldAndNonEmptyNewField_ShouldCreateCustomFieldValue`

File: docs/RELEASE_NOTES.md (hunk 5 above)
Comment (heshamMassoud) [reply docs/RELEASE_NOTES.md]: `buildNewOrModifiedCustomFieldsUpdateActions` is not an exposed utility so the users are actually not interesed about it.

What they are interested about is the called method and any exposed functionality that uses this:

```suggestion
    - **Commons** -  Fixed a bug in the custom fields update actions builders which generated duplicated unnecessary update actions for `null` custom field values. This affected any sync module where the resource contained custom fields (i.e. Product Sync, Category Sync, CartDiscount Sync and Inventory Sync). It also affected any update actions building utility in which the resource/sub-resource contained custom fields.  [#428](https://github.com/commercetools/commercetools-sync-java/issues/428)
```

File: docs/RELEASE_NOTES.md (hunk 4 above)
Comment (JudeNiroshan) [reply docs/RELEASE_NOTES.md]: 60bf8bbcef9744ef1cf8d94e1169d5bf64186889

File: docs/RELEASE_NOTES.md (hunk 5 above)
Comment (JudeNiroshan) [reply docs/RELEASE_NOTES.md]: 21c9a77337f4dd48d0cf7d59466d8f8388947db9

File: src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java (hunk 6 above)
Comment (JudeNiroshan) [reply src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: b378c4953ef11405bff4500c05b27e29b3cb4212

File: src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java (hunk 8 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java]: 89aab968ca658892ef582d74b6a2a5644592eb63

File: src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java (hunk 9 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtilsTest.java]: fda7ecdcdb2b57ad5fec294465ebb86088c9f5f3

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java (hunk 10 above)
Comment (JudeNiroshan) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: 11c58a79fddf5f35febca23ed818b71a4e2ff694

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java (hunk 11 above)
Comment (JudeNiroshan) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: cc77a94165412b5c689e3170647af8e9a5c2eb55

File: docs/RELEASE_NOTES.md (hunk 4 above)
Comment (heshamMassoud) [reply docs/RELEASE_NOTES.md]: It seems you didn't address this correctly 😢

File: docs/RELEASE_NOTES.md (hunk 5 above)
Comment (heshamMassoud) [reply docs/RELEASE_NOTES.md]: It seems you didn't address this correctly also 😢

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java (hunk 10 above)
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: but you still didn't add an integration test here.

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java (hunk 11 above)
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: thanks. Is there a case for sync_WithNullOldCustomFieldAndEmptyNewField_ShouldSetCustomField

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java [hunk 12]
```diff
@@ -586,6 +591,217 @@ void withMixedCasesOfPriceMatches_ShouldBuildActions() {
...
+
+        List<Price> priceList = productProjection.getMasterVariant().getPrices();
+        for (Price price : priceList) {
+            Set<String> fieldAsStringSet = price.getCustom().getFieldAsStringSet(SET_CUSTOM_FIELD_NAME);
+            assertThat(fieldAsStringSet).containsOnly("new something");
+        }
+    }
+
+    @Test
+    void sync_WithEmptyOldCustomFieldAndNullNewField_ShouldRemoveTheCustomField() {
```
Comment (heshamMassoud) [inline src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]:
```suggestion
    void sync_WithEmptyOldCustomFieldAndNullNewField_ShouldRemoveCustomField() {
```

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java [hunk 13]
```diff
@@ -586,6 +591,217 @@ void withMixedCasesOfPriceMatches_ShouldBuildActions() {
...
+
+        List<Price> priceList = productProjection.getMasterVariant().getPrices();
+        for (Price price : priceList) {
+            Set<String> fieldAsStringSet = price.getCustom().getFieldAsStringSet(SET_CUSTOM_FIELD_NAME);
+            assertThat(fieldAsStringSet).isEmpty();
+        }
+    }
+
+    @Test
+    void sync_WithNonEmptyOldCustomFieldAndNonEmptyNewField_ShouldUpdateCustomFieldSet() {
```
Comment (heshamMassoud) [inline src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]:
```suggestion
    void sync_WithNonEmptyOldCustomFieldAndNonEmptyNewField_ShouldSetCustomField() {
```

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java [hunk 14]
```diff
@@ -586,6 +591,217 @@ void withMixedCasesOfPriceMatches_ShouldBuildActions() {
         assertPricesAreEqual(productProjection.getMasterVariant().getPrices(), newPricesWithReferenceIds);
     }
 
+    @Test
+    void sync_WithEmptyOldCustomFieldAndNonEmptyNewField_ShouldCreateCustomFieldValue() {
```
Comment (heshamMassoud) [inline src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]:
```suggestion
    void sync_WithEmptyOldCustomFieldAndNonEmptyNewField_ShouldSetCustomFieldValue() {
```

File: docs/RELEASE_NOTES.md (hunk 4 above)
Comment (heshamMassoud) [reply docs/RELEASE_NOTES.md]: 1060c4a

File: docs/RELEASE_NOTES.md (hunk 5 above)
Comment (heshamMassoud) [reply docs/RELEASE_NOTES.md]: ac9311a

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java (hunk 14 above)
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: 699535a

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java (hunk 13 above)
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: 699535a

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java (hunk 12 above)
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: 699535a

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java (hunk 11 above)
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: 9f1bcf1

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java (hunk 10 above)
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithPricesIT.java]: 9f1bcf1
